### PR TITLE
Feat/#15 user authentication

### LINKED ISSUE #15: User Authentication
Implement ***User Authentication***  that allows the creation and connection of a user to our backend services using [JSON Web Tokens](https://jwt.io/) that provide access for a limited period of time.

### PR DESCRIPTION
**Pull request by MarianGiugioiu**
### This pull request solves issue:

Issue: #15 

### Which app is this issue related to?

- [x] REST API
- [ ] E-Learning React App
- [ ] ~~Admin Panel~~ - Not created yet

### Types of changes

- [ ] Bug fix (a non-breaking change which fixes an issue)
- [x] New feature (a non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

### Wiki Updates

- [ ] No need for updates
- [x] Requires an update ( _not done within this issue_ )
- [ ] Has been updated ( _done within this issue_ )

### Testing checklist

- [ ] Unit tests
- [ ] Functional tests
- [ ] E2E tests

#### Compatibility tests - **_For E-Learning React App Only_**

- [ ] Tested on Safari
- [ ] Tested on Chrome
- [ ] Tested on Mozilla Firefox

### What changes have been done within this issue?

I added Role endpoint.
I added User endpoint with the following extra methods:
- `loadByUsername` - overrides method from UserDetailsService and creates a UserDetails object based on the User entry with the provided username
-  `getOneByUsername` - returns an User entry with the provided username
- `getAllByRole` returns all User entries with the provided role name
- `register` creates a new User entry based on the provided RegisterForm object and assigns it the student role.

I added a service for generating and decoding jwt.
I added filters for auhentication and authorization processes.
I added configuration for security which handles aceess to the api resources based on the user roles.
I added dependencies for lombok, spring-security and java jwt in pom.xml.

### How should this be manually tested?

Firstly you need to create a new User entry using `/users/add` or `/users/register` paths and then you can use the `/users/login` path and provide the correct username and password to obtain the access and refresh tokens. 
In order to access any api resource accessible to the user, you need to combine the string `Bearer ` with the access token and put the result into the `Authorization` field of the request header.
When the access token expires, you can get a new one using the `/token/refresh` path and the refresh token.
Most api resources are only available to admin users and other users can only access and edit their own information.
For the `/users/username` path you need to use a query parameter named username.

### CODE REVIEW

File: apps/java-api/src/main/java/org/scoalaonline/api/DTO/RegisterForm.java [hunk 1]
```diff
@@ -0,0 +1,10 @@
+package org.scoalaonline.api.DTO;
+
+import lombok.Data;
+
+@Data
+public class RegisterForm {
+  private String Name;
```
Comment (CristiPV) [inline apps/java-api/src/main/java/org/scoalaonline/api/DTO/RegisterForm.java]: Change 'Name' to 'name' and make sure to change it in all the places where it is used.

File: apps/java-api/src/main/java/org/scoalaonline/api/controller/RoleController.java [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+@CrossOrigin
+@RestController
+@RequiredArgsConstructor
+@RequestMapping("/roles")
+public class RoleController {
+  private final RoleService roleService;
+
+  /**
+   * Sends HTTP status Response Entity with all the Role entries.
+   * @return a Response Entity with HTTP Status OK and a list of the User entries.
```
Comment (CristiPV) [inline apps/java-api/src/main/java/org/scoalaonline/api/controller/RoleController.java]:
```suggestion
   * @return a Response Entity with HTTP Status OK and a list of the Role entries.
```

File: apps/java-api/src/main/java/org/scoalaonline/api/controller/RoleController.java [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+  @PatchMapping( value = "/{id}" )
+  public ResponseEntity<Role> updateRole( @PathVariable( "id" ) String id, @RequestBody Role role ) {
+    Role updatedRole;
+    try
+    {
+      updatedRole = roleService.update( id, role );
+    } catch ( RoleNotFoundException e )
+    {
+      throw new ResponseStatusException( HttpStatus.NOT_FOUND, "PATCH: Role Not Found", e );
+    } catch ( RoleInvalidNameException e){
```
Comment (CristiPV) [inline apps/java-api/src/main/java/org/scoalaonline/api/controller/RoleController.java]:
```suggestion
    } catch ( RoleInvalidNameException e)
    {
```

Comment (CristiPV) [reply apps/java-api/src/main/java/org/scoalaonline/api/controller/RoleController.java]: Just to keep it the same as the one above.

File: apps/java-api/src/main/java/org/scoalaonline/api/controller/UserController.java [hunk 4]
```diff
@@ -0,0 +1,273 @@
...
+ GET:	"/users"	retrieves all the entries
+ GET:	"/users/{id}"	retrieves the entry with the provided id
+ GET:	"/users/username"	retrieves the entry with the provided username
+ GET:	"/users/role/{roleName}"	retrieves all the entries with the provided role
+ GET:	"/users/token/refresh"	offers new access token using the refresh token
+ POST: "/users"	creates a new entry
+ POST: "/users/register"	registers a new entry with the default role
+ PATCH:	"/users/{id}	edits the entry with the provided id
+ DELETE:	"/users/{id}	deletes the entry with the provided id
+
```
Comment (CristiPV) [inline apps/java-api/src/main/java/org/scoalaonline/api/controller/UserController.java]:
```suggestion
```

File: apps/java-api/src/main/java/org/scoalaonline/api/controller/UserController.java [hunk 5]
```diff
@@ -0,0 +1,273 @@
...
+    return new ResponseEntity<>(user, HttpStatus.OK);
+  }
+
+  /**
+   * Sends HTTP Response Entity with a specific User entry.
+   * Sends HTTP status Not Found if there is no entry with the provided username.
+   * Sends HTTP status Forbidden if the currently authenticated user is not the same
+   * as the one with the given username or does not have permission to see all User entries.
+   * @param username the username of the specific User.
+   * @param authentication information about the currently authenticated user.
```
Comment (CristiPV) [inline apps/java-api/src/main/java/org/scoalaonline/api/controller/UserController.java]:
```suggestion
   * @param username - the username of the specific User.
   * @param authentication - information about the currently authenticated user.
```

Comment (CristiPV) [reply apps/java-api/src/main/java/org/scoalaonline/api/controller/UserController.java]: For better readability.

File: apps/java-api/src/main/java/org/scoalaonline/api/controller/UserController.java [hunk 6]
```diff
@@ -0,0 +1,273 @@
...
+   * @return the Response Entity with a Status Code and a body.
+   */
+  @GetMapping(value = "/username")
+  @ResponseBody
```
Comment (CristiPV) [inline apps/java-api/src/main/java/org/scoalaonline/api/controller/UserController.java]:
```suggestion
  @ResponseBody
```
Is this needed ?

File: apps/java-api/src/main/java/org/scoalaonline/api/controller/UserController.java [hunk 7]
```diff
@@ -0,0 +1,273 @@
...
+  public ResponseEntity<User> getUserByUsername(@RequestParam String username, Authentication authentication) {
+    List<GrantedAuthority> authorities = new ArrayList<>();
```
Comment (CristiPV) [inline apps/java-api/src/main/java/org/scoalaonline/api/controller/UserController.java:96]:
```suggestion
    // Security Check
    List<GrantedAuthority> authorities = new ArrayList<>();
```

File: apps/java-api/src/main/java/org/scoalaonline/api/controller/UserController.java [hunk 8]
```diff
@@ -0,0 +1,273 @@
...
+   */
+  @PatchMapping( value = "/{id}" )
+  public ResponseEntity<User> updateUser( @PathVariable( "id" ) String id, @RequestBody User user ) {
+    User updatedUser;
+    try
+    {
+      updatedUser = userService.update( id, user );
+    } catch ( UserNotFoundException e )
+    {
+      throw new ResponseStatusException( HttpStatus.NOT_FOUND, "PATCH: User Not Found", e );
```
Comment (CristiPV) [inline apps/java-api/src/main/java/org/scoalaonline/api/controller/UserController.java]:
```suggestion
    } catch ( UserNotFoundException e ) {
      throw new ResponseStatusException( HttpStatus.NOT_FOUND, "PATCH: User Not Found", e );
```

Comment (CristiPV) [reply apps/java-api/src/main/java/org/scoalaonline/api/controller/UserController.java]: For consistency.

File: apps/java-api/src/main/java/org/scoalaonline/api/controller/UserController.java [hunk 9]
```diff
@@ -0,0 +1,273 @@
...
+    } catch ( UserInvalidNameException e){
```
Comment (CristiPV) [inline apps/java-api/src/main/java/org/scoalaonline/api/controller/UserController.java]:
```suggestion
    } catch ( UserInvalidNameException e) {
```

File: apps/java-api/src/main/java/org/scoalaonline/api/controller/UserController.java [hunk 10]
```diff
@@ -0,0 +1,273 @@
...
+      throw new ResponseStatusException( HttpStatus.BAD_REQUEST, "PATCH: User Invalid Name", e );
+    } catch (UserUsernameNotAllowedException e) {
+      throw new ResponseStatusException( HttpStatus.BAD_REQUEST, "PATCH: Can not change username", e );
```
Comment (CristiPV) [inline apps/java-api/src/main/java/org/scoalaonline/api/controller/UserController.java]:
```suggestion
      throw new ResponseStatusException( HttpStatus.BAD_REQUEST, "PATCH: Cannot change username", e );
```

File: apps/java-api/src/main/java/org/scoalaonline/api/service/RoleService.java [hunk 11]
```diff
@@ -0,0 +1,95 @@
...
+  @Override
+  public Role getOneById(String id) throws RoleNotFoundException {
+    return roleRepository.findById(id).orElseThrow(
+      () -> new RoleNotFoundException("Method getOneById: Role not found")
+    );
+  }
+
+  /**
+   * Adds a Role entry in the DB based on the received object.
+   * @param entry the Role entry.
```
Comment (CristiPV) [inline apps/java-api/src/main/java/org/scoalaonline/api/service/RoleService.java]:
```suggestion
   * @param entry - the Role entry.
```

Comment (CristiPV) [reply apps/java-api/src/main/java/org/scoalaonline/api/service/RoleService.java]: For clarity.

File: apps/java-api/src/main/java/org/scoalaonline/api/service/RoleService.java [hunk 12]
```diff
@@ -0,0 +1,95 @@
...
+      throw new RoleInvalidNameException("Method add: Name field can't be null.");
+
+    return roleRepository.save(roleToSave);
+  }
+
+  /**
+   * Updates the Role entry with the given id based on the received object.
+   * Throws an exception if no entry with that id was found.
+   * @param id - the id of the entry to update
+   * @param entry the Role entry.
```
Comment (CristiPV) [inline apps/java-api/src/main/java/org/scoalaonline/api/service/RoleService.java]:
```suggestion
   * @param entry - the Role entry.
```

File: apps/java-api/src/main/java/org/scoalaonline/api/service/RoleService.java [hunk 13]
```diff
@@ -0,0 +1,95 @@
...
+   */
+  @Override
+  public Role update(String id, Role entry) throws RoleNotFoundException, RoleInvalidNameException {
+    Role roleToUpdate = roleRepository.findById(id).orElseThrow(
+      () -> new RoleNotFoundException("Method update: Role not found")
+    );
+    if(entry.getName() != null && !entry.getName().equals("")) {
+      roleToUpdate.setName(entry.getName());
+    } else {
+      throw new RoleInvalidNameException("Method update: Name Field Can't Be Null");
```
Comment (CristiPV) [inline apps/java-api/src/main/java/org/scoalaonline/api/service/RoleService.java]:
```suggestion
      throw new RoleInvalidNameException("Method update: Name field can't be null.");
```

Comment (CristiPV) [reply apps/java-api/src/main/java/org/scoalaonline/api/service/RoleService.java]: This way it matches the one from `add` method.

File: apps/java-api/src/main/java/org/scoalaonline/api/service/RoleService.java [hunk 14]
```diff
@@ -0,0 +1,95 @@
...
+   * entry with that id can be found
+   * @param id - id of the Role entry
+   * @throws RoleNotFoundException
+   */
+  @Override
+  public void delete(String id) throws RoleNotFoundException {
+    if(roleRepository.findById(id).isPresent())
+      roleRepository.deleteById(id);
+    else
+      throw new RoleNotFoundException("Method delete: Role Not Found");
```
Comment (CristiPV) [inline apps/java-api/src/main/java/org/scoalaonline/api/service/RoleService.java]:
```suggestion
      throw new RoleNotFoundException("Method delete: Role not found.");
```

File: apps/java-api/src/main/java/org/scoalaonline/api/service/UserService.java [hunk 15]
```diff
@@ -0,0 +1,267 @@
...
+   * @return UserDetails object
+   * @throws UsernameNotFoundException
+   */
+  @Override
+  public UserDetails loadUserByUsername(String username) throws UsernameNotFoundException {
+    User user = userRepository.findByUsername(username).orElseThrow(
+      () -> new UsernameNotFoundException("Method loadUserByUsername: User not found")
+    );
+
+    log.info("User {} found in te database", username);
```
Comment (CristiPV) [inline apps/java-api/src/main/java/org/scoalaonline/api/service/UserService.java]:
```suggestion
    log.info("User {} found in the database", username);
```

File: apps/java-api/src/main/java/org/scoalaonline/api/service/UserService.java [hunk 16]
```diff
@@ -0,0 +1,267 @@
...
+
+  /**
+   * Retrieves one User entry with the given username from the DB
+   * or throws an error if no entry with that username is found.
+   * @param username - username of the User entry
+   * @return the User entry
+   * @throws UserNotFoundException
+   */
+  public User getOneByUsername(String username) throws UserNotFoundException{
+    log.info("Fetching user {}", username);
```
Comment (CristiPV) [inline apps/java-api/src/main/java/org/scoalaonline/api/service/UserService.java]: Would be nice to add these logs throughout the entire class.

File: apps/java-api/src/main/java/org/scoalaonline/api/service/UserService.java [hunk 17]
```diff
@@ -0,0 +1,267 @@
...
+
+  /**
+   * Retrieves a list of all User entries found in the DB who have the Role given by roleName
+   * @param roleName - name of the Role entry
+   * @throws RoleNotFoundException
+   * @return the list of User entries
+   */
+  public List<User> getAllByRole(String roleName) {
+    List<User> users = userRepository.findAllByRolesContaining(roleName);
+    return users;
```
Comment (CristiPV) [inline apps/java-api/src/main/java/org/scoalaonline/api/service/UserService.java]:
```suggestion
    return userRepository.findAllByRolesContaining(roleName);
```

File: apps/java-api/src/main/java/org/scoalaonline/api/service/UserService.java [hunk 18]
```diff
@@ -0,0 +1,267 @@
...
+  }
+
+  /**
+   * Adds a User entry in the DB based on the received object.
+   * @param entry the User entry.
```
Comment (CristiPV) [inline apps/java-api/src/main/java/org/scoalaonline/api/service/UserService.java]:
```suggestion
   * @param entry - the User entry.
```

File: apps/java-api/src/main/java/org/scoalaonline/api/service/UserService.java [hunk 19]
```diff
@@ -0,0 +1,267 @@
...
+    else
+      throw new UserInvalidRolesException("Method add: Roles field can't be null.");
+
+    return userRepository.save(userToSave);
+  }
+
+  /**
+   * Adds a User entry in the DB based on the received RegisterForm.
+   * Adds ROLE_STUDENT to the list of roles
+   * @param entry the User entry.
```
Comment (CristiPV) [inline apps/java-api/src/main/java/org/scoalaonline/api/service/UserService.java]:
```suggestion
   * @param entry - the User entry.
```

File: apps/java-api/src/main/java/org/scoalaonline/api/service/UserService.java [hunk 20]
```diff
@@ -0,0 +1,267 @@
...
+    addRoleToUser(userToSave, DEFAULT_ROLE);
+
+    return userRepository.save(userToSave);
+  }
+
+  /**
+   * Updates the User entry with the given id based on the received object.
+   * Throws an exception if no entry with that id was found.
+   * @param id - the id of the entry to update
+   * @param entry the User entry.
```
Comment (CristiPV) [inline apps/java-api/src/main/java/org/scoalaonline/api/service/UserService.java]:
```suggestion
   * @param entry - the User entry.
```

File: apps/java-api/src/main/java/org/scoalaonline/api/service/UserService.java [hunk 21]
```diff
@@ -0,0 +1,267 @@
...
+    UserInvalidPasswordException,
+    UserInvalidRolesException{
+    User userToUpdate = userRepository.findById(id).orElseThrow(
+      () -> new UserNotFoundException("Method update: User not found")
+    );
+
+    if(entry.getName() != null && !entry.getName().equals("")) {
+      userToUpdate.setName(entry.getName());
+    } else {
+      throw new UserInvalidNameException("Method update: Name Field Can't Be Null");
```
Comment (CristiPV) [inline apps/java-api/src/main/java/org/scoalaonline/api/service/UserService.java]:
```suggestion
      throw new UserInvalidNameException("Method update: Name field can't be null.");
```

File: apps/java-api/src/main/java/org/scoalaonline/api/service/UserService.java [hunk 22]
```diff
@@ -0,0 +1,267 @@
...
+    }
+
+    log.info(String.valueOf(userToUpdate.getUsername().equals(entry.getUsername())));
```
Comment (CristiPV) [inline apps/java-api/src/main/java/org/scoalaonline/api/service/UserService.java]:
```suggestion
```

File: apps/java-api/src/main/java/org/scoalaonline/api/service/UserService.java [hunk 23]
```diff
@@ -0,0 +1,267 @@
...
+    if(entry.getUsername() != null && !entry.getUsername().equals("") && !userToUpdate.getUsername().equals(entry.getUsername()))
+      throw new UserUsernameNotAllowedException("Method update: Can not change username.");
```
Comment (CristiPV) [inline apps/java-api/src/main/java/org/scoalaonline/api/service/UserService.java]:
```suggestion
      throw new UserUsernameNotAllowedException("Method update: Cannot change username.");
```

File: apps/java-api/src/main/java/org/scoalaonline/api/service/UserService.java [hunk 24]
```diff
@@ -0,0 +1,267 @@
...
+   * entry with that id can be found
+   * @param id - id of the User entry
+   * @throws UserNotFoundException
+   */
+  @Override
+  public void delete(String id) throws UserNotFoundException {
+    if(userRepository.findById(id).isPresent())
+      userRepository.deleteById(id);
+    else
+      throw new UserNotFoundException("Method delete: User Not Found");
```
Comment (CristiPV) [inline apps/java-api/src/main/java/org/scoalaonline/api/service/UserService.java]:
```suggestion
      throw new UserNotFoundException("Method delete: User not found.");
```

File: apps/java-api/src/main/java/org/scoalaonline/api/service/UserService.java [hunk 25]
```diff
@@ -0,0 +1,267 @@
...
+  }
+
+  /**
+   * Retrieves one Role entry with the given roleName from the DB
+   * Add the Role to User entry given by user
```
Comment (CristiPV) [inline apps/java-api/src/main/java/org/scoalaonline/api/service/UserService.java]:
```suggestion
   * Adds the Role to User entry given by user
```

File: apps/java-api/src/main/java/org/scoalaonline/api/service/UserService.java [hunk 26]
```diff
@@ -0,0 +1,267 @@
...
+   * @param user - User entry
+   * @param roleName - name of the Role entry
+   * @throws RoleNotFoundException
+   */
+  public void addRoleToUser(User user, String roleName) throws RoleNotFoundException {
+    Role role = roleRepository.findByName(roleName).orElseThrow(
+      () -> new RoleNotFoundException("Method addRoleToUser: Role not found")
+    );
+    log.info("Adding role {} to user {}", role.getName(), user.getName());
+    user.getRoles().add(role);
```
Comment (CristiPV) [inline apps/java-api/src/main/java/org/scoalaonline/api/service/UserService.java]: Make custom `getter` & `setter` for this one to add encapsulation.

File: apps/java-api/src/main/java/org/scoalaonline/api/service/JWTService.java [hunk 27]
```diff
@@ -0,0 +1,66 @@
...
+/**
+ * Contains the logic for managing JWT
+ */
+@Service
+@RequiredArgsConstructor
+@Slf4j
+public class JWTService {
+  private Algorithm algorithm = Algorithm.HMAC256("secret".getBytes());
+  private Date access_timer =  new Date(System.currentTimeMillis() + 24 * 60 * 60 * 1000);
+  private Date refresh_timer =  new Date(System.currentTimeMillis() + 24 * 60 * 60 * 1000);
```
Comment (CristiPV) [inline apps/java-api/src/main/java/org/scoalaonline/api/service/JWTService.java]:
```suggestion
  private Date accessTimer =  new Date(System.currentTimeMillis() + 24 * 60 * 60 * 1000);
  private Date refreshTimer =  new Date(System.currentTimeMillis() + 24 * 60 * 60 * 1000);
```
Refactor these variables to use *camelCase*.

File: apps/java-api/src/main/java/org/scoalaonline/api/service/JWTService.java [hunk 28]
```diff
@@ -0,0 +1,66 @@
...
+    return JWT.create()
+      .withSubject(username)
+      .withExpiresAt(access_timer)
+      .withIssuer(requestURL)
+      .sign(algorithm);
+  }
+
+  /**
+   * Generates a decoded JWT with informations about the authenticated user
+   * ased on the given token, using the algorithm
```
Comment (CristiPV) [inline apps/java-api/src/main/java/org/scoalaonline/api/service/JWTService.java]:
```suggestion
   * based on the given token, using the algorithm
```

File: apps/java-api/src/main/java/org/scoalaonline/api/service/JWTService.java [hunk 29]
```diff
@@ -0,0 +1,66 @@
...
+  public String getRefreshToken(String username, String requestURL) {
+    return JWT.create()
+      .withSubject(username)
+      .withExpiresAt(access_timer)
+      .withIssuer(requestURL)
+      .sign(algorithm);
+  }
+
+  /**
+   * Generates a decoded JWT with informations about the authenticated user
```
Comment (CristiPV) [inline apps/java-api/src/main/java/org/scoalaonline/api/service/JWTService.java]:
```suggestion
   * Generates a decoded JWT with information on the authenticated user
```

File: apps/java-api/src/main/java/org/scoalaonline/api/security/UserSecurity.java [hunk 30]
```diff
@@ -0,0 +1,32 @@
...
+ */
+@Component("userSecurity")
+@RequiredArgsConstructor
+public class UserSecurity {
+  private final UserService userService;
+
+  /**
+   * Checks if the user with the specified id is the same as the currently authenticated user
+   * @param authentication information about the currently authenticated user.
+   * @param id the id of the specific user
```
Comment (CristiPV) [inline apps/java-api/src/main/java/org/scoalaonline/api/security/UserSecurity.java]:
```suggestion
   * @param id - the id of the specific user
```

File: apps/java-api/src/main/java/org/scoalaonline/api/filter/CustomAuthenticationFilter.java [hunk 31]
```diff
@@ -0,0 +1,112 @@
...
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+import static org.springframework.http.HttpStatus.FORBIDDEN;
+import static org.springframework.util.MimeTypeUtils.APPLICATION_JSON_VALUE;
+
+/**
+ *Creates a filter for the process of authentication
```
Comment (CristiPV) [inline apps/java-api/src/main/java/org/scoalaonline/api/filter/CustomAuthenticationFilter.java]:
```suggestion
 * Creates a filter for the process of authentication
```

File: apps/java-api/src/main/java/org/scoalaonline/api/security/SecurityConfig.java [hunk 32]
```diff
@@ -0,0 +1,94 @@
...
+    http.csrf().disable();
+    http.sessionManagement().sessionCreationPolicy(SessionCreationPolicy.STATELESS);
+
+    http.authorizeRequests().antMatchers("/users/login/**", "/users/token/refresh/**").permitAll();
+
+    http.authorizeRequests().antMatchers("/roles").hasAnyAuthority("ROLE_ADMIN");
+
+    http.authorizeRequests().antMatchers(HttpMethod.GET, "/users").hasAnyAuthority("ROLE_ADMIN");
+    http.authorizeRequests().antMatchers(HttpMethod.GET, "/users/role/**").hasAnyAuthority("ROLE_ADMIN");
+    http.authorizeRequests().antMatchers(HttpMethod.GET, "/users/username").permitAll();
```
Comment (CristiPV) [inline apps/java-api/src/main/java/org/scoalaonline/api/security/SecurityConfig.java]:
```suggestion
    // Custom authorization implemented in UserController.
    http.authorizeRequests().antMatchers(HttpMethod.GET, "/users/username").permitAll();
```

**Review by CristiPV — CHANGES_REQUESTED** [review]

File: apps/java-api/src/main/java/org/scoalaonline/api/security/SecurityConfig.java [hunk 33]
```diff
@@ -0,0 +1,94 @@
...
+  /**
+   * Adds filters for Authentication and Authorization processes
+   * Configures access to API resources
+   * @param http
+   * @throws Exception
+   */
+  @Override
+  protected void configure(HttpSecurity http) throws Exception {
+    CustomAuthenticationFilter customAuthenticationFilter = new CustomAuthenticationFilter(authenticationManagerBean(),jwtService);
+    customAuthenticationFilter.setFilterProcessesUrl("/users/login");
```
Comment (CristiPV) [inline apps/java-api/src/main/java/org/scoalaonline/api/security/SecurityConfig.java:50]: Add a custom endpoint for logging out as well.

**Review by CristiPV — CHANGES_REQUESTED** [review]

File: apps/java-api/src/main/java/org/scoalaonline/api/controller/UserController.java (hunk 6 above)
Comment (MarianGiugioiu) [reply apps/java-api/src/main/java/org/scoalaonline/api/controller/UserController.java]: No, it was deleted.

File: apps/java-api/src/main/java/org/scoalaonline/api/DTO/RegisterForm.java (hunk 1 above)
Comment (MarianGiugioiu) [reply apps/java-api/src/main/java/org/scoalaonline/api/DTO/RegisterForm.java]: Done. The getters and setters had the same name for both 'Name' and 'name'.

File: apps/java-api/src/main/java/org/scoalaonline/api/filter/CustomAuthenticationFilter.java [hunk 34]
```diff
@@ -0,0 +1,112 @@
...
+    try {
+      Map<String, String> requestMap = new ObjectMapper().readValue(request.getInputStream(), Map.class);
+      username = requestMap.get("username");
+      password = requestMap.get("password");
+    } catch (IOException e) {
+      throw new AuthenticationServiceException(e.getMessage(), e);
+    }
+
+    log.info("Username is {}", username);
+    log.info("Password is {}", password);
```
Comment (CristiPV) [inline apps/java-api/src/main/java/org/scoalaonline/api/filter/CustomAuthenticationFilter.java]:
```suggestion
    log.info("Password is {}", password);
```

Comment (CristiPV) [reply apps/java-api/src/main/java/org/scoalaonline/api/filter/CustomAuthenticationFilter.java]: Password should not be logged.

File: apps/java-api/src/main/java/org/scoalaonline/api/service/UserService.java [hunk 35]
```diff
@@ -0,0 +1,310 @@
...
+@RequiredArgsConstructor
+public class UserService implements ServiceInterface<User>, UserDetailsService {
+  private static final String DEFAULT_ROLE = "ROLE_STUDENT";
+
+  private final UserRepository userRepository;
+  private final RoleRepository roleRepository;
+  private final PasswordEncoder passwordEncoder;
+
+  /**
+   * Retrieves User entry with given username
```
Comment (CristiPV) [inline apps/java-api/src/main/java/org/scoalaonline/api/service/UserService.java]:
```suggestion
   * Retrieves User entry with given the username
```

File: apps/java-api/src/main/java/org/scoalaonline/api/service/UserService.java [hunk 36]
```diff
@@ -0,0 +1,310 @@
...
+   * Creates Spring Security UserDetails based on User entry
```
Comment (CristiPV) [inline apps/java-api/src/main/java/org/scoalaonline/api/service/UserService.java]:
```suggestion
   * Creates Spring Security UserDetails based on the User entry
```

File: apps/java-api/src/main/java/org/scoalaonline/api/service/UserService.java [hunk 37]
```diff
@@ -0,0 +1,310 @@
...
+   * @param username
+   * @return UserDetails object
+   * @throws UsernameNotFoundException
+   */
+  @Override
+  public UserDetails loadUserByUsername(String username) throws UsernameNotFoundException {
+    User user = userRepository.findByUsername(username).orElseThrow(
+      () -> new UsernameNotFoundException("Method loadUserByUsername: User not found.")
+    );
```
Comment (CristiPV) [inline apps/java-api/src/main/java/org/scoalaonline/api/service/UserService.java]:
```suggestion
    User user = userRepository.findByUsername(username).orElseThrow(
      () -> {
        log.error("Username not found.");
        return new UsernameNotFoundException("Method loadUserByUsername: Username not found.");
      }
    );
```

File: apps/java-api/src/main/java/org/scoalaonline/api/service/UserService.java [hunk 38]
```diff
@@ -0,0 +1,310 @@
...
+   * @param id - id of the User entry
+   * @throws UserNotFoundException
+   */
+  @Override
+  public void delete(String id) throws UserNotFoundException {
+    if(userRepository.findById(id).isPresent()) {
+      log.info("Deleting user with id {}...", id);
+      userRepository.deleteById(id);
+    } else {
+      log.error("User not found in te database.");
```
Comment (CristiPV) [inline apps/java-api/src/main/java/org/scoalaonline/api/service/UserService.java]:
```suggestion
      log.error("User not found in the database.");
```

File: apps/java-api/src/main/java/org/scoalaonline/api/service/RoleService.java [hunk 39]
```diff
@@ -0,0 +1,117 @@
...
+   * @throws RoleNotFoundException
+   */
+  @Override
+  public void delete(String id) throws RoleNotFoundException {
+    if(roleRepository.findById(id).isPresent()) {
+      log.info("Deleting role with id {}...", id);
+      roleRepository.deleteById(id);
+    }
+    else {
+      log.error("Role not found in te database.");
```
Comment (CristiPV) [inline apps/java-api/src/main/java/org/scoalaonline/api/service/RoleService.java]:
```suggestion
      log.error("Role not found in the database.");
```

File: apps/java-api/src/main/java/org/scoalaonline/api/controller/RoleController.java [hunk 40]
```diff
@@ -0,0 +1,118 @@
...
+      throw new ResponseStatusException( HttpStatus.BAD_REQUEST, "PATCH: Role Invalid Name", e );
+    }
+
+    return new ResponseEntity<>( updatedRole, HttpStatus.OK );
+  }
+
+  /**
+   * Deletes a role entry and sends a Response Entity
+   * with the Status OK or Not Found if there is no entry with the provided id
+   * @param id the id of the Role to be deleted.
```
Comment (CristiPV) [inline apps/java-api/src/main/java/org/scoalaonline/api/controller/RoleController.java]:
```suggestion
   * @param id - the id of the Role to be deleted.
```

File: apps/java-api/src/main/java/org/scoalaonline/api/controller/RoleController.java [hunk 41]
```diff
@@ -0,0 +1,118 @@
...
+      throw new ResponseStatusException( HttpStatus.BAD_REQUEST, "POST: Role Invalid Name", e );
+    }
+    return new ResponseEntity<>(savedRole, HttpStatus.CREATED);
+  }
+
+  /**
+   * Sends HTTP status Response Entity with the Role entry that has been updated.
+   * Sends HTTP status Not Found if the Role cannot be found.
+   * Sends HTTP status Invalid Value if the Role to be posted is invalid.
+   * @param id the id of the Role to be updated.
```
Comment (CristiPV) [inline apps/java-api/src/main/java/org/scoalaonline/api/controller/RoleController.java]:
```suggestion
   * @param id - the id of the Role to be updated.
```

File: apps/java-api/src/main/java/org/scoalaonline/api/controller/RoleController.java [hunk 42]
```diff
@@ -0,0 +1,118 @@
...
+    {
+      throw new ResponseStatusException( HttpStatus.NOT_FOUND, "GET: Role Not Found", e );
+    }
+    return new ResponseEntity<>(role, HttpStatus.OK);
+  }
+
+  /**
+   * Sends HTTP status Response Entity with the Role entry that has been created.
+   * Sends HTTP status Invalid value if the Role to be posted is invalid.
+   * @param role the Role to be added in the db.
```
Comment (CristiPV) [inline apps/java-api/src/main/java/org/scoalaonline/api/controller/RoleController.java]:
```suggestion
   * @param role - the Role to be added in the db.
```

File: apps/java-api/src/main/java/org/scoalaonline/api/controller/RoleController.java [hunk 43]
```diff
@@ -0,0 +1,118 @@
...
+  @GetMapping(value = {"", "/"})
+  public ResponseEntity<List<Role>> getAllRoles () {
+    List<Role> roles = roleService.getAll();
+    return new ResponseEntity<>(roles, HttpStatus.OK);
+  }
+
+  /**
+   * Sends an HTTP Response Entity with a specific Role entry
+   * Sends Status OK or Status Not Found if there is no entry with the provided id
+   * @param id the id of the specific Role
```
Comment (CristiPV) [inline apps/java-api/src/main/java/org/scoalaonline/api/controller/RoleController.java]:
```suggestion
   * @param id - the id of the specific Role
```

**Review by CristiPV — CHANGES_REQUESTED** [review]

**Review by CristiPV — APPROVED** [review]

**Pull request merged**